predicates/source/source_test.go

```diff
@@ -305,3 +313,104 @@ func TestMatchingClientIP(t *testing.T) {
 		})
 	}
 }
+
+var random = rand.New(rand.NewSource(time.Now().UnixNano()))
+
+func generateIPCidr() string {
+	if m := random.Int(); m%10 == 0 {
+		n := random.Intn(16) + 16 // /16 .. /32
+		return fmt.Sprintf("%d.%d.%d.%d/%d", random.Intn(256), random.Intn(256), random.Intn(256), random.Intn(256), n)
+	} else if m%2 == 0 {
+		n := random.Intn(16) + 48 // /16 .. /32
+		return fmt.Sprintf("%02x:%02x::af:%02x:%02x/%d", random.Intn(16), random.Intn(16), random.Intn(16), random.Intn(16), n)
+	}
+	return fmt.Sprintf("%d.%d.%d.%d", random.Intn(256), random.Intn(256), random.Intn(256), random.Intn(256))
+}
+
+func generateIPCidrStrings(n int) []string {
+	a := make([]string, 0)
+	for i := 0; i < n; i++ {
+		a = append(a, generateIPCidr())
+	}
+	return a
+}
+
+func BenchmarkClientIP10(b *testing.B) {
+	benchSource(b, 10, NewClientIP())
+}
+func BenchmarkClientIP100(b *testing.B) {
+	benchSource(b, 100, NewClientIP())
+}
+func BenchmarkClientIP1k(b *testing.B) {
+	benchSource(b, 1_000, NewClientIP())
+}
+func BenchmarkClientIP10k(b *testing.B) {
+	benchSource(b, 10_000, NewClientIP())
+}
+func BenchmarkClientIP100k(b *testing.B) {
+	benchSource(b, 100_000, NewClientIP())
+}
+
+func BenchmarkSource10(b *testing.B) {
+	benchSource(b, 10, New())
+}
+func BenchmarkSource100(b *testing.B) {
+	benchSource(b, 100, New())
+}
+func BenchmarkSource1k(b *testing.B) {
+	benchSource(b, 1_000, New())
+}
+func BenchmarkSource10k(b *testing.B) {
+	benchSource(b, 10_000, New())
+}
+func BenchmarkSource100k(b *testing.B) {
+	benchSource(b, 100_000, New())
+}
+
+func BenchmarkSourceFromLast10(b *testing.B) {
+	benchSource(b, 10, NewFromLast())
+}
+func BenchmarkSourceFromLast100(b *testing.B) {
+	benchSource(b, 100, NewFromLast())
+}
+func BenchmarkSourceFromLast1k(b *testing.B) {
+	benchSource(b, 1_000, NewFromLast())
+}
+func BenchmarkSourceFromLast10k(b *testing.B) {
+	benchSource(b, 10_000, NewFromLast())
+}
+func BenchmarkSourceFromLast100k(b *testing.B) {
+	benchSource(b, 100_000, NewFromLast())
+}
+
+func stringsToArgs(a []string) []interface{} {
+	res := make([]interface{}, 0, len(a))
+	for _, s := range a {
+		var e any = s
+		res = append(res, e)
+	}
+	return res
+}
+
+func benchSource(b *testing.B, k int, spec routing.PredicateSpec) {
+	target := "195.5.1.21"
+	xff := "195.5.1.23, 195.5.1.24"
+	a := generateIPCidrStrings(k)
+	args := stringsToArgs(a)
+
+	pred, err := spec.Create(args)
+	if err != nil {
+		b.Fatalf("Failed to create %s with args '%v': %v", spec.Name(), args, err)
+	}
+	req := &http.Request{
+		RemoteAddr: target + ":12345",
```

Comment:
The actual error was not to have a port added here, which was fixed in https://github.com/zalando/skipper/pull/2195/commits/fac4152d925420f92fdb25da5a07f83ccd73612e.
So the test was testing an ignored error of 
		h, _, _ := net.SplitHostPort(r.RemoteAddr)
that will make h=="" and was input to
 src, _ = netip.ParseAddr(h) // new
 src = net.ParseIP(h) // old

Reply:
type Request struct {
    // RemoteAddr allows HTTP servers and other software to record the network address
    // that sent the request, usually for logging. This field is not filled in by
    // ReadRequest and has no defined format. The HTTP server in this package sets
    // RemoteAddr to an "IP:port" address before invoking a handler. This field is
    // ignored by the HTTP client.
    RemoteAddr string